crates/net/network/src/transactions/fetcher.rs

```diff
@@ -215,27 +216,27 @@ impl<N: NetworkPrimitives> TransactionFetcher<N> {
     pub fn find_any_idle_fallback_peer_for_any_pending_hash(
         &mut self,
         hashes_to_request: &mut RequestTxHashes,
-        is_session_active: impl Fn(&PeerId) -> bool,
+        peers: &HashMap<PeerId, PeerMetadata<N>>,
         mut budget: Option<usize>, // search fallback peers for max `budget` lru pending hashes
     ) -> Option<PeerId> {
         let mut hashes_pending_fetch_iter = self.hashes_pending_fetch.iter();
 
         let idle_peer = loop {
+            if let Some(ref mut bud) = budget {
+                *bud = bud.saturating_sub(1);
+                if *bud == 0 {
+                    return None
+                }
+            }
```

Comment:
why is this moved to top of loop?

Reply:
Early exit if there's no budget, no need to go trough all the logic if there's no budget.

Reply:
Reverted for now.